reframe/frontend/dependency.py

```diff
@@ -79,3 +80,44 @@ def print_deps(graph):
 
 def validate_deps(graph):
     """Validate dependency graph."""
+
+    # Reduce test case graph to a test name only graph; this disallows
+    # pseudo-dependencies as follows:
+    #
+    # (t0, e1) -> (t1, e1)
+    # (t1, e0) -> (t0, e0)
+    #
+    # This reduction step will result in a graph description with duplicate
+    # entries in the adjacency list; this is not a problem, cos they will be
+    # filtered out during the DFS traversal below.
+    test_graph = {}
+    for case, deps in graph.items():
+        test_deps = [d.check.name for d in deps]
+        try:
+            test_graph[case.check.name] += test_deps
+        except KeyError:
+            test_graph[case.check.name] = test_deps
+
+    # Check for cyclic dependencies in the test name graph
+    visited = set()
+    unvisited = list(
+        itertools.zip_longest(test_graph.keys(), [], fillvalue=None)
+    )
+    path = []
+    while unvisited:
+        node, parent = unvisited.pop()
+        while path and path[-1] != parent:
+            path.pop()
+
+        adjacent = reversed(test_graph[node])
+        path.append(node)
+        for n in adjacent:
+            if n in path:
+                cycle_str = '->'.join(path + [n])
+                raise DependencyError(
+                    'found cyclic dependency between tests: ' + cycle_str)
+
+            if n not in visited:
+                unvisited.append((n, node))
+
+        visited.add(node)
```

Comment:
Here you either have to move also `node` from the unvisited, or add an if block on line 108 to continue if the node is already in the visited ones.

Reply:
`node` is popped out from the `unvisited` at the beginning of the loop (`pop()` removes the element):

```python
        node, parent = unvisited.pop()
```

If I had to check if node inside `unvisited` is already visited, then there is a semantic problem.